File: services/cart-service/src/cart/domain/money.ts
import {
  CurrencyMismatchException,
  InvalidCartOperationException,
} from './cart.exceptions';

export class Money {
  private constructor(
    private readonly _amount: number,
    private readonly _currency: string,
  ) {}

  static of(amount: number, currency = 'VND'): Money {
    if (amount < 0) {
      throw new InvalidCartOperationException(
        'Amount must be greater than or equal to 0',
      );
    }

    return new Money(Math.round(amount), currency);
  }

  static zero(currency = 'VND'): Money {
    return new Money(0, currency);
  }

  get amount(): number {
    return this._amount;
  }

  get currency(): string {
    return this._currency;
  }

  add(other: Money): Money {
    if (this._currency !== other._currency) {
      throw new CurrencyMismatchException(this._currency, other._currency);
    }

    return new Money(this._amount + other._amount, this._currency);
  }

  multiply(multiplier: number): Money {
    return new Money(this._amount * multiplier, this._currency);
  }

  equals(other: Money): boolean {
    return this._amount === other._amount && this._currency === other._currency;
  }
}

File: services/order-service/src/main/java/com/vnshop/orderservice/domain/Money.java
package com.vnshop.orderservice.domain;

import java.math.BigDecimal;
import java.util.Objects;

public record Money(BigDecimal amount, String currency) {
    public static final Money ZERO = new Money(BigDecimal.ZERO);

    public Money(BigDecimal amount) {
        this(amount, "VND");
    }

    public Money {
        Objects.requireNonNull(amount, "amount is required");
        currency = currency == null || currency.isBlank() ? "VND" : currency;
        if (!"VND".equals(currency)) {
            throw new IllegalArgumentException("currency must be VND");
        }
        if (amount.signum() < 0) {
            throw new IllegalArgumentException("amount cannot be negative");
        }
        if (amount.scale() != 0) {
            throw new IllegalArgumentException("VND amount cannot have decimals");
        }
    }

    public Money add(Money other) {
        Objects.requireNonNull(other, "other is required");
        return new Money(amount.add(other.amount), currency);
    }

    public Money subtract(Money other) {
        Objects.requireNonNull(other, "other is required");
        return new Money(amount.subtract(other.amount), currency);
    }
}
